SetupDataPkg/Library/ConfigVariableListLib/ConfigVariableListLib.c

```diff
@@ -0,0 +1,292 @@
+    // know how many entries there are. We are returning full list of entries.
+    *ConfigVarListPtr = NULL;
+    *ConfigVarListPtr = AllocatePool (BinSize);
+  }
+
+  if (!*ConfigVarListPtr) {
+    DEBUG ((DEBUG_ERROR, "%a Failed to allocate memory for ConfigVarListPtr\n", __FUNCTION__));
+    return EFI_OUT_OF_RESOURCES;
+  }
+
+  while (ListIndex < BinSize) {
+    // index into variable list
+    VarList = (CONFIG_VAR_LIST_HDR *)((CHAR8 *)BlobPtr + ListIndex);
+
+    if (ListIndex + sizeof (*VarList) + VarList->NameSize + VarList->DataSize + sizeof (*Guid) +
+        sizeof (Attributes) + sizeof (CRC32) > BinSize)
+    {
+      // the NameSize and DataInBinSize have bad values and are pushing us past the end of the binary
+      DEBUG ((DEBUG_ERROR, "%a Configuration VarList had bad NameSize or DataInBinSize\n", __FUNCTION__));
+      ASSERT (FALSE);
+      Status = EFI_COMPROMISED_DATA;
+      break;
+    }
+
+    /*
+     * Var List is in DmpStore format:
+     *
+     *  struct {
+     *    CONFIG_VAR_LIST_HDR VarList;
+     *    CHAR16 Name[VarList->NameSize/2];
+     *    EFI_GUID Guid;
+     *    UINT32 Attributes;
+     *    CHAR8 DataInBin[VarList-DataSize];
+     *    UINT32 CRC32; // CRC32 of all bytes from VarList to end of DataInBin
+     *  }
+     */
+    NameInBin  = (CHAR16 *)(VarList + 1);
+    Guid       = (EFI_GUID *)((CHAR8 *)NameInBin + VarList->NameSize);
+    Attributes = *(UINT32 *)(Guid + 1);
+    DataInBin  = (CHAR8 *)Guid + sizeof (*Guid) + sizeof (Attributes);
+    CRC32      = *(UINT32 *)(DataInBin + VarList->DataSize);
+
+    // on next iteration, skip past this variable
+    ListIndex += sizeof (*VarList) + VarList->NameSize + VarList->DataSize + sizeof (*Guid) + sizeof (Attributes)
+                 + sizeof (CRC32);
+
+    VarName = AllocatePool (VarList->NameSize);
+    if (VarName == NULL) {
+      DEBUG ((DEBUG_ERROR, "%a Failed to allocate memory for VarName size: %u\n", __FUNCTION__, VarList->NameSize));
+      Status = EFI_OUT_OF_RESOURCES;
+      goto Exit;
+    }
+
+    CopyMem (VarName, NameInBin, VarList->NameSize);
+
+    // This check needs to come after we allocate VarName so we can compare against a 16 bit aligned
+    // string and not assert
+    if (ConfigVarName && (0 != StrnCmp (ConfigVarName, VarName, VarList->NameSize / 2))) {
+      // Not the entry we are looking for
+      // While we are here, set the Name entry to NULL so we can tell if we found the entry later
+      (*ConfigVarListPtr)->Name = NULL;
+      FreePool (VarName);
+      continue;
+    }
+
+    Data = AllocatePool (VarList->DataSize);
+    if (Data == NULL) {
+      DEBUG ((DEBUG_ERROR, "%a Failed to allocate memory for Data size: %u\n", __FUNCTION__, VarList->DataSize));
+      Status = EFI_OUT_OF_RESOURCES;
+      // Free VarName here, as other memory gets freed in exit routine, but we do not increment
+      // *ConfigVarListCount for this entry
+      FreePool (VarName);
+      goto Exit;

```

Comment:
There has been gotos and returns in the same function. Let's make it more consistent (probably 1 entry and 1 exit).